src/transform-palette.ts

```diff
@@ -0,0 +1,104 @@
+import {
+    ADDRESS_CLAMP_TO_EDGE,
+    PIXELFORMAT_RGBA32F,
+    GraphicsDevice,
+    Mat4,
+    Texture
+} from 'playcanvas';
+
+// mapping from Mat4 to transposed 3x4 matrix
+const idx = [
+    0, 4, 8, 12,
+    1, 5, 9, 13,
+    2, 6, 10, 14
+];
+
+// texture data stores 512 matrices per row: 512 * 3 * 4 (rgba) floats
+const width = 512 * 3;
+
+// wraps a palette of transform data. transforms are stored as 3x4 (non-perspective)
+// matrices
+class TransformPalette {
+    getTransform: (index: number, transform: Mat4) => void;
+    setTransform: (index: number, transform: Mat4) => void;
+    alloc: (num?: number) => number;
+    free: (num?: number) => void;
+    texture: Texture;
+
+    constructor(device: GraphicsDevice, initialSize = 4096) {
+        let texture: Texture;
+        let data: Float32Array;
+
+        // reallocate the storage texture and copy over old data
+        const realloc = (width: number, height: number) => {
+            const newTexture = new Texture(device, {
```

Comment:
just a note that when you move this to the engine v2 you can use texture.resize()